src/coreclr/jit/hwintrinsicxarch.cpp

```diff
@@ -1415,15 +1415,192 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
 
         case NI_Vector128_ConvertToDouble:
         case NI_Vector256_ConvertToDouble:
+        case NI_Vector512_ConvertToDouble:
+        {
+            assert(sig->numArgs == 1);
+            assert(varTypeIsLong(simdBaseType) || simdBaseType == TYP_FLOAT);
+            if (IsBaselineVector512IsaSupportedOpportunistically())
+            {
+                if (varTypeIsLong(simdBaseType))
+                {
+                    intrinsic = (simdSize == 16) ? NI_AVX512DQ_VL_ConvertToVector128Double
+                              : (simdSize == 32) ? NI_AVX512DQ_VL_ConvertToVector256Double
+                              : NI_AVX512DQ_ConvertToVector512Double;
+                }
+                else
+                {
+                    intrinsic = (simdSize == 16) ? NI_SSE2_ConvertToVector128Double
+                              : (simdSize == 32) ? NI_AVX_ConvertToVector256Double
+                              : NI_AVX512F_ConvertToVector512Double;
+                }
+
+                op1     = impSIMDPopStack();
+                retNode = gtNewSimdHWIntrinsicNode(retType, op1, intrinsic, simdBaseJitType, simdSize);
+            }
+            break;
+        }
+
         case NI_Vector128_ConvertToInt64:
         case NI_Vector256_ConvertToInt64:
+        case NI_Vector512_ConvertToInt64:
+        {
+            assert(sig->numArgs == 1);
+            assert(simdBaseType == TYP_DOUBLE);
+#ifdef TARGET_AMD64
+            if (IsBaselineVector512IsaSupportedOpportunistically())
+            {
```

Comment:
This could probably do with a comment giving the equivalent C# code and a brief explanation.

Something like:
```csharp
// Replace QNaN and SNaN with Zero
op1 = Avx512F.Fixup(op1, op1, Vector128.Create<long>(0x88), 0);

// Convert from double to long, replacing any values that were greater than or equal to MaxValue with MaxValue
// Values that were less than or equal to MinValue will already be MinValue
return Vector128.ConditionalSelect(
    Vector128.LessThan(op1, Vector128.Create<double>(long.MaxValue)).AsInt64(),
    Avx512DQ.VL.ConvertToVector128Int64(op1),
    Vector128.Create<long>(long.MaxValue)
);
```

Reply:
Same would ideally be done for other new cases below.

Reply:
Sure.